components/drivetrain.py

```diff
@@ -1,21 +1,180 @@
+
+        motion_profile = MotionProfile.generate_motion_profile(
+            acceleration_time=1,
+            deceleration_time=1,
+            max_speed=1,
+            target_distance=distance)
+
+        coefs = PIDCoefficients(p=1.0, i=0.0, d=0.0)
+
+        # Set current position to zero
+        self._reset_encoder_position()
+
+        self.profile_executor = ProfileExecutor(
+            coefs, motion_profile,
+            lambda: self._get_encoder_position() / 1023,
+            lambda output: self.forward_at(-output), 0.08)
+
+    def backward(self, feet=0, inches=0, meters=0):
+        """Use a motion profile and PID control to efficiently
+         move the robot the specified distance backward"""
+        self.forward(-feet, -inches, -meters)
+
+    def rotate(self, degrees=0):
+        """Use a motion profile and PID control to efficiently
+         turn the robot the number of degrees - positive is clockwise,
+         negative is counter-clockwise"""
+        radians = degrees * (math.pi / 180)
+
+        # When called with the same arguments, update executor
+        if radians == self.profile_arguments:
+            return self._update_executor()
+
+        # When target is switched, issue warning
+        if (self.profile_arguments is not None and
+                radians != self.profile_arguments):
+            print("Use Drivetrain.cancel_motion_profile to change profile!",
+                  "Switching to new profile...")
+        self.profile_arguments = radians
+
+        motion_profile = MotionProfile.generate_motion_profile(
+            acceleration_time=1,
+            deceleration_time=1,
+            max_speed=0.6,
+            target_distance=radians)
+
+        self._zero_gyro()
+
+        coefs = PIDCoefficients(p=1.0, i=0.1, d=0.0)
+        self.profile_executor = ProfileExecutor(
+            coefs, motion_profile,
+            lambda: self._get_gyro_angle(),
+            lambda output: self.turn_at(output), 0.001
+        )
+
+    def cancel_motion_profile(self):
+        self.profile_executor = None
+        self.profile_arguments = None
+
+    def profile_in_use(self):
+        return self.profile_executor is not None
+
     def execute(self):
-        self.robot_drive.arcadeDrive(self.forward, self.rotation)
+        self.robot_drive.arcadeDrive(self.forward_speed, self.rotation)
 
         self.rotation = 0
-        self.forward = 0
+        self.forward_speed = 0
+
+    def _reset_encoder_position(self):
+        self.fl_motor.setEncPosition(0)
+
+    def _get_encoder_position(self):
+        return -self.fl_motor.getPosition()
+
+    def _zero_gyro(self):
+        self.gyro_offset = self.gyro.getAngle()
+
+    def _get_gyro_angle(self):
+        return (self.gyro.getAngle() - self.gyro_offset) * (math.pi / 180.0)
+
+    def _update_executor(self):
+        executor_finished = self.profile_executor.update()
+        if executor_finished:
+            return True
```

Comment:
Can we cancel the executor here?

Reply:
That’s what I did originally, but it forces the consumer to stop calling forward or rotate immediately or a new profile will be started. Should I cancel there anyway?